### PR TITLE
Remove ksonnet and use new API version

### PR DESCRIPTION
**Pull request by scottilee**
Related to: #3721 notebook controller E2E test

This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kubeflow/4171)

### CODE REVIEW

**Comment by scottilee** [comment]
@jlewi Just want verify that I can't do `util.run(["kubectl", "apply", "-f", "jupyter_test.yaml"], cwd=app_dir)`? I guess I have to use the API?

**Comment by scottilee** [comment]
Used the python kubernetes client instead and ended up with this error: `HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Notebook in version \"v1beta1\" cannot be handled as a Deployment: no kind \"Notebook\" is registered for version \"kubeflow.org/v1beta1\" in scheme \"k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:29\"","reason":"BadRequest","code":400}`

File: kubeflow/jupyter/tests/jupyter_test.py [hunk 1]
```diff
@@ -112,19 +112,21 @@ def test_jupyter(env, namespace):
...
+    resp = k8s_apps_v1.create_namespaced_deployment(
+      body=wf_result, namespace=namespace)
+    logging.info("Deployment created: status='%s'" % resp.metadata.name)
 
-  name = "jupyter-test"
-  service = "jupyter-test"
-  component = "jupyter"
-  params = ""
-  ks_util.setup_ks_app(app_dir, env, namespace, component, params)
+  # this_dir = os.path.dirname(__file__)
```
Comment (jlewi) [inline kubeflow/jupyter/tests/jupyter_test.py]: Delete these commented out lines

File: kubeflow/jupyter/tests/jupyter_test.py [hunk 2]
```diff
@@ -112,19 +112,21 @@ def test_jupyter(env, namespace):
...
+  # app_dir = os.path.join(this_dir, "test_app")
 
-  util.run([ks_cmd, "apply", env, "-c", component], cwd=app_dir)
```
Comment (jlewi) [inline kubeflow/jupyter/tests/jupyter_test.py]: Can we delete the old jsonnet file?

Comment (scottilee) [reply kubeflow/jupyter/tests/jupyter_test.py]: Yes, I removed `jupyter_test.jsonnet`, `sync-notebook_test.jsonnet`, and `notebooks_test.jsonnet`. Let me know if you meant just the `jupyter_test.jsonnet`.

File: kubeflow/jupyter/tests/jupyter_test.py (hunk 1 above)
Comment (scottilee) [reply kubeflow/jupyter/tests/jupyter_test.py]: Done.

**Comment by jlewi** [comment]
Using kfctl should work but using the API is better.

Not sure why the tests are failing. Latest failure is

```
util.py                     71 INFO     go: github.com/kubeflow/kubeflow/components/profile-controller/v2@v2.0.0: unknown revision components/profile-con
troller/v2.0.0
```

Which looks unrelated to this PR.

Can you try rebasing on master?

**Comment by jlewi** [comment]
@scottilee Do you know what the latest test failures were?

/cc @Tabrizian

**Comment by nrchakradhar** [comment]
I had a look into the failures. All the failures are related to Notebook/v1beta1 is not a registered kind.

https://github.com/kubeflow/kubeflow/blob/0137d254964b56f1891d23127915ee56d8e91fa3/kubeflow/jupyter/tests/jupyter_test.py#L37-L38

v1beta1 is present in the manifest but the failure is still seen.

https://github.com/kubeflow/manifests/blob/609b0dda6d6124e0893685f7972de83d156ea0dd/jupyter/notebook-controller/base/crd.yaml#L15-L20

HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},**"status":"Failure"**,"message":"Notebook
 in **version \"v1beta1\"** cannot be handled as a Deployment: no kind **\"Notebook\"** is registered for version \"kubeflow.or
g/v1beta1\" in scheme \"k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:29\"","reason":"BadRequest","code":400}

File: kubeflow/jupyter/tests/jupyter_test.py [hunk 3]
```diff
@@ -112,19 +112,17 @@ def test_jupyter(env, namespace):
...
-  params = ""
-  ks_util.setup_ks_app(app_dir, env, namespace, component, params)
 
-  util.run([ks_cmd, "apply", env, "-c", component], cwd=app_dir)
+  with open("jupyter_test.yaml") as params: 
+    wf_result = yaml.load(params)
+    namespace = wf_result['metadata']['namespace']
+    name = wf_result['metadata']['name']
+    k8s_apps_v1 = k8s_client.AppsV1Api()
+    resp = k8s_apps_v1.create_namespaced_deployment(
```
Comment (jlewi) [inline kubeflow/jupyter/tests/jupyter_test.py]: This code doesn't look right. Why are you creating a deployment here? The  contents of jupyter_test.yaml is a Notebook so you need to use the custom resource APi.

You can take a look at how we handle Argo workflows
https://github.com/kubeflow/testing/blob/f1d5dcd63b91f9b327beaf1e3aa48543f7c00ad5/py/kubeflow/testing/run_e2e_workflow.py#L357
We need to do something similar.

**Comment by jlewi** [comment]
Thanks @nrchakradhar 

@scottilee I left a couple comments on the PR about where I think the bug might be.

Have you tried deploying your own Kubeflow cluster and then running the test against that cluster? That would probably allow you to develop and test faster.

**Comment by nrchakradhar** [comment]
@scottilee the latest failure is due to namespace being xgboost. I guess the older tests were running in "kubeflow" namespace.

HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"namespac
es \"xboost-demo\" not found","reason":"NotFound","details":{"name":"xboost-demo","kind":"namespaces"},"code":404

https://github.com/kubeflow/kubeflow/blob/1980561d3e7f326220e46eaeffec8d43a7e4af9b/kubeflow/jupyter/tests/jupyter_test.yaml#L8

https://github.com/kubeflow/kubeflow/blob/1980561d3e7f326220e46eaeffec8d43a7e4af9b/kubeflow/jupyter/tests/jupyter_test.py#L119

wf_result['metadata']['namespace'] = namespace from passed arguments.

https://github.com/kubeflow/kubeflow/blob/1980561d3e7f326220e46eaeffec8d43a7e4af9b/kubeflow/jupyter/tests/jupyter_test.py#L100

**Comment by scottilee** [comment]
@nrchakradhar thanks! I have updated it. How are you debugging these errors? I don't see those error outputs in the presubmit logs.

**Comment by nrchakradhar** [comment]
The latest failure seems to be from 
https://github.com/kubeflow/kubeflow/blob/32186454c63a9c72f95dc3fb5e277f16bb25d4d8/kubeflow/jupyter/tests/jupyter_test.py#L129
It might be better to follow like:
https://github.com/kubeflow/testing/blob/f1d5dcd63b91f9b327beaf1e3aa48543f7c00ad5/py/kubeflow/testing/run_e2e_workflow.py#L365
which logs everything than just one field.

**Comment by jlewi** [comment]
@nrchakradhar @scottilee 

Argo fetches logs from the pods so if the pods are GC'd then the logs are no longer available.

At that point you can look in stackdriver for the logs.
More info here
https://github.com/kubeflow/testing#debugging-failed-tests

**Comment by nrchakradhar** [comment]
Gentle ping @scottilee

**Comment by scottilee** [comment]
Okay, I see `HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the namespace of the provided object does not match the namespace sent on the request","reason":"BadRequest","code":400}`.  Trying a fix with updating line 132 to `ns` instead.

**Comment by scottilee** [comment]
Hm, not sure what the failure is about, seems to not like this line? `results = util.wait_for_cr_condition(api_client, GROUP, PLURAL, VERSION, ns, name, conditions)`

File: kubeflow/jupyter/tests/jupyter_test.yaml [hunk 4]
```diff
@@ -0,0 +1,34 @@
+apiVersion: kubeflow.org/v1beta1
+kind: Notebook
+metadata:
+  labels:
+    addgcpsecret: "true"
+    app: xgboost-demo
+  name: xgboost-demo
+  namespace: kubeflow
```
[the remark this answers is not in the document]
Comment (jlewi) [reply kubeflow/jupyter/tests/jupyter_test.yaml]: +1

File: kubeflow/jupyter/tests/jupyter_test.yaml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+spec:
+  template:
+    spec:
+      containers:
+      - env: []
+        image: gcr.io/kubeflow-images-public/tensorflow-1.13.1-notebook-cpu:v0.5.0
```
Comment (jlewi) [inline kubeflow/jupyter/tests/jupyter_test.yaml]: You probably want to use one of the newer images.

**Comment by jlewi** [comment]
@scottilee My suggestion would be to create your own Kubeflow cluster and then run jupyter_test.py against that. Then you will be able to inspect the resources and see what's happening.

**Comment by nrchakradhar** [comment]
@jlewi @scottilee I successfully validated the util.wait_for_cr_condition() for jupyter notebook in a local cluster which had v1alpha1 version. I believe this should work for v1beta1 also.
@scottilee Let me know if you want any further help on this for submitting the PR.

**Comment by scottilee** [comment]
@nrchakradhar thanks for checking that, so are you saying that my changes should work? Do you know what I need to do to get the test to pass?

**Comment by nrchakradhar** [comment]
@scottilee I guess so. Can you change the image as suggested by @jlewi and resubmit for testing. Also, there is an option to pass callbacks to the wait_for_cr_condition which can be used to get back the responses as done in tf-operator https://github.com/kubeflow/tf-operator/blob/d5d783cbd7710b4b9fc631e83cc270082913c3f1/py/kubeflow/tf_operator/simple_tfjob_tests.py#L54

**Comment by jlewi** [comment]
@scottilee have you tried running jupyter_test.py against your own Kubeflow cluster? That would make it easier to debug. If you need access to a GCP project let me know I can give you access to one.

**Comment by scottilee** [comment]
@jlewi yes, that would be very helpful!

**Comment by jlewi** [comment]
@scottilee actually since you are working on KCC setup and have access to the folder for KCC projects you should be able to create a project in that folder using KCC and then deploy KF into it for testing.

You can also submit a PR to add yourself to the devrel team
https://github.com/kubeflow/internal-acls/blob/master/devrel-team.members.txt

**Comment by scottilee** [comment]
@nrchakradhar any interest in taking this over?

**Comment by nrchakradhar** [comment]
@scottilee Unfortunately due to some restrictions from my employer,  I cannot contribute currently. Sorry.

**Comment by jlewi** [comment]
@scottilee do you need someone else to pick this up?

**Comment by scottilee** [comment]
Yes, I think it's time to pass this on, haha.

**Comment by jlewi** [comment]
It looks like @gabrielwen fixed this in kubeflow/kfctl#142.

Closing this PR.

**Pull request closed without merging**